pkg/async/schedule/flytescheduler/workflow_executor.go

```diff
@@ -0,0 +1,251 @@
+		if !*s.Active {
+			continue
+		} else {
+			fromTime = s.UpdatedAt
+		}
+		err := w.CatchUpSingleSchedule(ctx, s, fromTime, toTime)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (w *workflowExecutor) CatchUpSingleSchedule(ctx context.Context, s models.SchedulableEntity, fromTime time.Time, toTime time.Time) error {
+	nameOfSchedule := w.goGfInterface.GetScheduleName(s)
+	var catchUpTimes []time.Time
+	var err error
+	catchUpTimes, err = w.goGfInterface.GetCatchUpTimes(s, fromTime, toTime)
+	if err != nil {
+		return err
+	}
+	var catchupTime time.Time
+	for _, catchupTime = range catchUpTimes {
+		_ = w.rateLimiter.Take()
+		err := fire(ctx, w.executionManager, catchupTime, s)
+		if err != nil {
+			logger.Errorf(ctx, "unable to fire the schedule %v at %v time due to %v", s.Name, catchupTime, err)
+			return err
+		} else {
+			w.snapshot.UpdateLastExecutionTime(nameOfSchedule, catchupTime)
+		}
+	}
+
+	return nil
+}
+
+func fire(ctx context.Context, executionManager mgInterfaces.ExecutionInterface, scheduledTime time.Time,
+	s models.SchedulableEntity) error {
+
+	literalsInputMap := map[string]*core.Literal{}
+	literalsInputMap[s.KickoffTimeInputArg] = &core.Literal{
+		Value: &core.Literal_Scalar{
+			Scalar: &core.Scalar{
+				Value: &core.Scalar_Primitive{
+					Primitive: &core.Primitive{
+						Value: &core.Primitive_Datetime{
+							Datetime: timestamppb.New(scheduledTime),
+						},
+					},
+				},
+			},
+		},
+	}
+
+	executionRequest := admin.ExecutionCreateRequest{
+		Project: s.Project,
+		Domain:  s.Domain,
+		Name:    "f" + strings.ReplaceAll(uuid.New().String(), "-", "")[:19],
+		Spec: &admin.ExecutionSpec{
+			LaunchPlan: &core.Identifier{
+				ResourceType: core.ResourceType_LAUNCH_PLAN,
+				Project:      s.Project,
+				Domain:       s.Domain,
+				Name:         s.Name,
+				Version:      s.Version,
+			},
+			Metadata: &admin.ExecutionMetadata{
+				Mode:        admin.ExecutionMetadata_SCHEDULED,
+				ScheduledAt: timestamppb.New(scheduledTime),
+			},
+			// No dynamic notifications are configured either.
+		},
+		// No additional inputs beyond the to-be-filled-out kickoff time arg are specified.
+		Inputs: &core.LiteralMap{
+			Literals: literalsInputMap,
+		},
+	}
+	if !*s.Active {
+		// no longer active
+		logger.Debugf(ctx, "schedule %v is no longer active", s.Name)
+		return nil
+	}
+	_, err := executionManager.CreateExecution(context.Background(), executionRequest, scheduledTime)
+	if err != nil {
+		logger.Error(ctx, "failed to create execution create request %v due to %v", executionRequest, err)
+		return err
+	}
+	return nil
+}
+
+func (w *workflowExecutor) Run() {
+	ctx := context.Background()
+	c, err := w.db.SchedulableEntityRepo().GetAll(ctx)
+	if err != nil {
+		panic(fmt.Errorf("unable to run the workflow executor after reading the schedules due to %v", err))
+	}
+	// Start the go routine to write the snapshot map to the configured snapshot file
+	go w.CheckPointState(ctx)
```

Comment:
can we handle, error and panic here. this go routine may error out or have a panic this should be handled. This can be done with some helper methods from k8s apiachinery too

Reply:
I do not think we should start the checkpointer, till we catchup?

Reply:
I have moved the checkpointer after catchup, the reason i kept it before is so that everytime catch up runs it makes some progress in case we have continuous crashes or stop/restart. 

Eg : If the catch was to be done for a 1 week , then checkpointing before can be useful. Let me know what you think. 1 week doesn't sounds reasonable so i am ok with having the checkpointer after the catchup